django/db/models/fields/__init__.py

```diff
@@ -794,7 +794,11 @@ def db_type(self, connection):
         # exactly which wacky database column type you want to use.
         data = self.db_type_parameters(connection)
         try:
-            return connection.data_types[self.get_internal_type()] % data
+            column_type = connection.data_types[self.get_internal_type()]
+            # column_type is either a single-parameter function or a string
+            if callable(column_type):
+                return column_type(data)
+            return column_type % data
```

Comment:
So I guess this comes down to whether to allow callables here, vs (say) a custom field (in `contrib.postgres`?) 

@charettes @felixxm — can I ask your opinion there please?

Reply:
I don't think that returning a `callable` is a huge deal but it can be avoided by having `DatabaseWrapper.data_types` entries turned instances of a `str` subclass that implements `__mod__`

```python
class Varchar(str):
    def __new__(cls):
        return super().__new__(cls, "varchar(%(max_length)s)")

    def __mod__(self, data):
        if data["max_length"] is None:
            return "varchar"
        return super().__mod__(data)
```

```python
>>> vc = Varchar()
'varchar(%(max_length))'
>>> isinstance(vc, str)
True
>>> vc % {'max_length': 255}
'varchar(355)'
>>> vc % {'max_length': None}
'varchar'

Reply:
That's fair enough, I guess it depends on whether we esteem that there might be more cases like this in the future or not -- if it is the case I would prefer callables, otherwise Simon's suggestion is fine for a one-time thing.